wdl-grammar/src/tree/dive.rs

```diff
@@ -0,0 +1,259 @@
+//! Utilities for traversing a syntax tree while collecting elements of interest
+//! (i.e., "diving" for elements).
+
+use std::collections::VecDeque;
+use std::iter::FusedIterator;
+use std::marker::PhantomData;
+
+use rowan::Language;
+
+use crate::SyntaxElement;
+use crate::SyntaxNode;
+use crate::WorkflowDescriptionLanguage;
+
+/// An iterator that explores every node within a
+/// [`SyntaxNode`](rowan::SyntaxNode) yielding all of the elements match the
+/// conditions laid out in `match_predicate`.
+///
+/// There are also facilities to ignore certain subtrees within the tree via the
+/// `ignore_predicate` callback. If the entire tree is desired, one can simply
+/// have the `ignore_predicate` always return `false`.
+#[derive(Debug)]
+pub struct DiveIterator<L, M, I>
+where
+    L: Language,
+    M: Fn(&rowan::SyntaxElement<L>) -> bool,
+    I: Fn(&rowan::SyntaxNode<L>) -> bool,
+{
+    /// The queue of nodes to be visited.
+    nodes: VecDeque<rowan::SyntaxElement<L>>,
+    /// The function that evaluates when checking if an element matches.
+    match_predicate: M,
+    /// The function that evaluates when checking if the children underneath a
+    /// node in the tree should be ignored.
+    ignore_predicate: I,
+    /// A pointer to the [`Language`].
+    _phantom: PhantomData<L>,
+}
+
+impl<L, M, I> DiveIterator<L, M, I>
+where
+    L: Language,
+    M: Fn(&rowan::SyntaxElement<L>) -> bool,
+    I: Fn(&rowan::SyntaxNode<L>) -> bool,
+{
+    /// Creates a new [`DiveIterator`].
+    ///
+    /// Note that this iterator traverses elements as a breadth-first traversal
+    /// (otherwise known as level-order).
+    pub fn new(root: rowan::SyntaxElement<L>, match_predicate: M, ignore_predicate: I) -> Self {
+        let mut nodes = VecDeque::new();
+        nodes.push_back(root);
+
+        Self {
+            nodes,
+            match_predicate,
+            ignore_predicate,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<L, M, I> Iterator for DiveIterator<L, M, I>
+where
+    L: Language,
+    M: Fn(&rowan::SyntaxElement<L>) -> bool,
+    I: Fn(&rowan::SyntaxNode<L>) -> bool,
+{
+    type Item = rowan::SyntaxElement<L>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some(element) = self.nodes.pop_front() {
+            match &element {
+                rowan::SyntaxElement::Node(node) => {
+                    if !(self.ignore_predicate)(node) {
+                        self.nodes.extend(node.children_with_tokens());
+                    }
+                }
+                // NOTE: tokens have no children to explore.
+                rowan::SyntaxElement::Token(_) => {}
+            }
+
+            if (self.match_predicate)(&element) {
```

Comment:
At first glance this feels like a job for `Iterator::filter`.

I.e. this iterator always yields every element in the BFS (excluding subtrees that are ignored) and one can adapt the iterator with `filter` to exclude the element from the result.

Related: is the `clone` necessary? Seems like `element` is owned here and could just be returned, but that's just running the source through my brain's borrow-checker.

Reply:
Okay, I've addressed both of these.